common/src/api/internal/shared.rs

```diff
@@ -704,33 +704,130 @@ pub struct ResolvedVpcRouteSet {
 
 /// Describes the purpose of the dataset.
 #[derive(
-    Debug, Serialize, Deserialize, JsonSchema, Clone, Copy, PartialEq, Eq,
+    Debug,
+    Serialize,
+    Deserialize,
+    JsonSchema,
+    Clone,
+    PartialEq,
+    Eq,
+    Ord,
+    PartialOrd,
+    Hash,
 )]
-#[serde(rename_all = "snake_case")]
+#[serde(tag = "type", rename_all = "snake_case")]
 pub enum DatasetKind {
-    Crucible,
+    // Durable datasets for zones
+
+    // This renaming exists for backwards compatibility -- this enum variant
+    // was serialized to "all-zones-request" as "cockroach_db" and should
+    // stay that way, unless we perform an explicit schema change.
+    #[serde(rename = "cockroach_db")]
     Cockroach,
+    Crucible,
     Clickhouse,
     ClickhouseKeeper,
     ExternalDns,
     InternalDns,
+
+    // Zone filesystems
+    ZoneRoot,
+    Zone {
+        name: String,
+    },
+
+    // Other datasets
+    Debug,
 }
 
+impl DatasetKind {
+    pub fn dataset_should_be_encrypted(&self) -> bool {
+        match self {
+            // We encrypt all datasets except Crucible.
+            //
+            // Crucible already performs encryption internally, and we
+            // avoid double-encryption.
+            DatasetKind::Crucible => false,
+            _ => true,
+        }
+    }
+
+    /// Returns true if this dataset is delegated to a non-global zone.
+    pub fn zoned(&self) -> bool {
+        use DatasetKind::*;
+        match self {
+            Cockroach | Crucible | Clickhouse | ClickhouseKeeper
+            | ExternalDns | InternalDns => true,
+            ZoneRoot | Zone { .. } | Debug => false,
+        }
+    }
+
+    /// Returns the zone name, if this is dataset for a zone filesystem.
+    ///
+    /// Otherwise, returns "None".
+    pub fn zone_name(&self) -> Option<String> {
+        if let DatasetKind::Zone { name } = self {
+            Some(name.clone())
+        } else {
+            None
+        }
+    }
+}
+
+// Be cautious updating this implementation:
+//
+// - It should align with [DatasetKind::FromStr], below
+// - The strings here are used here comprise the dataset name, stored durably
+// on-disk
 impl fmt::Display for DatasetKind {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         use DatasetKind::*;
         let s = match self {
             Crucible => "crucible",
-            Cockroach => "cockroach",
+            Cockroach => "cockroachdb",
```

Comment:
When is the `Display` implementation used and when is a value serialized? I'm wondering why this variant is `"cockroachdb"` here but `"cockroach_db"` when serialized.

Reply:
+1, i would have a slight preference for the `Display` and `Serialize` strings being identical...

Reply:
+1 as well. For `ZoneKind`, which is related, we already have 4 different string representations X_X: https://github.com/oxidecomputer/omicron/blob/812c713a40c5c6a8e340583eb505b12c533c55bf/nexus-sled-agent-shared/src/inventory.rs#L380

Reply:
So my mishaps with naming here were **intended** to keep backwards compatibility between:

- Names for cockroach we have in deployed systems (e.g., the `zfs list`-ed name is `cockroachdb`)
- Names for cockroach we have stored on-disk in configuration files (e.g., grepping for `cockroach_db` in the `schema` directory of Omicron shows that it's used in e.g., `all-zones-requests.json`, which is used for bootstrapping "what zones get auto-launched when we reboot")

**To be clear, I think I misinterpreted this in my original PR**.

I was attempting to merge `DatasetKind` (from this file) and `DatasetType` (from `sled-storage/src/dataset.rs`) without breaking backwards compatibility. It was in the `all-zones-requests.json` configuration file where I saw the underscored "cockroach_db" name and tried to keep compatibility.

HOWEVER, upon closer inspection, the `cockroach_db` name actually comes from `OmicronZoneType` in `nexus-sled-agent-shared/src/inventory.rs`, so I **should** be able to just completely stick with the `cockroachdb` name (no underscores) in this `DatasetKind` variant.

Updated in 93134c25232d244cd3c54dcbd6d847640f9525ab